File: src/main/java/io/bindernews/thegrackle/actions/AddHitsAction.java
package io.bindernews.thegrackle.actions;

import com.megacrit.cardcrawl.actions.AbstractGameAction;
import com.megacrit.cardcrawl.cards.AbstractCard;
import com.megacrit.cardcrawl.cards.CardGroup;
import com.megacrit.cardcrawl.dungeons.AbstractDungeon;
import io.bindernews.thegrackle.variables.ExtraHitsVariable;
import lombok.val;

import java.util.Arrays;
import java.util.function.Consumer;

public class AddHitsAction extends AbstractGameAction {

    private final AbstractCard card;
    private final int amount;
    private final CardGroup[] cardGroups;

    public AddHitsAction(AbstractCard card, int amount, CardGroup[] cardGroups) {
        this.card = card;
        this.amount = amount;
        this.cardGroups = cardGroups;
    }

    @Override
    public void update() {
        val clz = card.getClass();
        val hv = ExtraHitsVariable.inst;

        Consumer<AbstractCard> applyToCard = card -> {
            hv.setBaseValue(card, hv.baseValue(card) + amount);
            card.applyPowers();
        };

        applyToCard.accept(card);
        Arrays.stream(cardGroups)
                .flatMap(a -> a.group.stream())
                .parallel()
                .filter(clz::isInstance)
                .forEach(applyToCard);
        isDone = true;
    }

    public static CardGroup[] getPlayerCardGroups() {
        val p = AbstractDungeon.player;
        return new CardGroup[]{p.hand, p.discardPile, p.exhaustPile, p.drawPile};
    }
}

File: src/main/kotlin/io/bindernews/thegrackle/actions/AddHitsAction.kt
package io.bindernews.thegrackle.actions

import com.megacrit.cardcrawl.actions.AbstractGameAction
import com.megacrit.cardcrawl.cards.AbstractCard
import com.megacrit.cardcrawl.cards.CardGroup
import com.megacrit.cardcrawl.dungeons.AbstractDungeon
import io.bindernews.thegrackle.helper.baseExtraHits
import java.util.*

class AddHitsAction(
    private val card: AbstractCard, amount: Int, private val cardGroups: Array<CardGroup>
) : AbstractGameAction() {
    init {
        this.amount = amount
    }

    override fun update() {
        val clz: Class<out AbstractCard> = card.javaClass
        card.baseExtraHits += amount
        card.applyPowers()
        Arrays.stream(cardGroups)
            .flatMap { a: CardGroup -> a.group.stream() }
            .parallel()
            .filter { clz.isInstance(it) }
            .forEach {
                it.baseExtraHits += amount
                it.applyPowers()
            }
        isDone = true
    }

    companion object {
        fun getPlayerCardGroups(): Array<CardGroup> {
            val p = AbstractDungeon.player
            return arrayOf(p.hand, p.discardPile, p.exhaustPile, p.drawPile)
        }
    }
}